### ISSUE TITLE
Events firehose

### ISSUE THREAD

**Issue by domwong** [m3o/dev#347]
To gain insight in to what's happening in our platform we need an event firehose. 

The basic premise is, everytime *something* interesting happens you create a json blob (event) that encapsulates what happened and you publish it. This stream of events is your firehose. You then have two consumers:
1. an events aggregation service which basically stores ALL the events and allows for querying
2. anything that might be interested and wants to do async processing. Example, you could have an history service which listens for "service.created|deleted|updated" events and persists them to a DB to be queried and provide a history of events for a service

An event should have at the very least:
- timestamp - unix timestamp
- name - e.g, "service.created"

Things to consider
- should there be a well known repository for event definitions? Events are really part of an API and there should be some contract which defines the event and its fields

Speak to Chris and Dom while designing

**Comment by domwong** [comment]
Other things to consider
- what storage is backing this? Feels like cockroach won't be the right thing
- how does querying work? Ideally you want to do things like "show me all the events for this customer" , "show me all events for this namespace"

**Comment by asim** [comment]
If its event streaming does this mean we need to run an equivalent of kafka? 

In prior iterations we've used a message bus like nsq and stored in cassandara, so like event sourcing. It's a pattern that every service ends up needing so if we can support event streaming/querying and we're the first users thats pretty good. My thoughts were nats + cockroach would be good but you're right, overloading this database for that purpose is wrong.

NATS has a counterpart called nats-streaming for persistence which might be useful for us
- https://github.com/nats-io/nats-streaming-server
- https://github.com/nats-io/jetstream
- https://docs.nats.io/nats-streaming-concepts/intro

**Comment by domwong** [comment]
> If its event streaming does this mean we need to run an equivalent of kafka?

Probably at some point yes. We can build it without the persistence part but we need to be damn sure our consumer is up all the time or we'll miss messages. Maybe for now we're OK with missing a message or two during beta? Guess it depends on how long it would take to implement vs without persistence.

I guess we can do some experimenting with different technologies and as long as we build it in such a way that storage and the pub sub system are pluggable then we don't necessarily back ourselves in to a corner.

**Comment by asim** [comment]
Yes, this comes back to go-micro and micro missing an "events" or "stream" interface. Many people tried to add durability, acking, etc to the go-micro/broker but that never felt like the right place for it. If we defined a go-micro/events or go-micro/stream interface then we could use it for this exact purpose.

**Comment by chrusty** [comment]
Having been in this situation many times before, I strongly suggest that we consider the following:

1) Don't just stick everything onto one topic. This is almost completely useless (it means that a consumer needs to filter messages, which all still need to be delivered from the messaging system to the servce instance). It will be more useful to have a topic for each event type. Options are:
    - Publish to one topic then have another service which reads all of those and splits them to their own topics.
    - Publish messages to their own topics (one per event-type). This is light-weight and removes the dependency on having an "event splitter", but it means that we can't do useful things like dumping every message on the firehose to cloud storage without subscribing to all of the different topics.
    - Publish each event twice (once to a "firehose" topic, and again to the specific event-type topic. This is quite easy, but does expose us to the "what if the first one succeeds but the second one fails" scenario.

2) Schema library is an absolute must. I usually define a "header" schema which is imported by default in every event. The header contains enough metadata to serve our generic routing and observability requirements:
    - event_type
    - timestamp
    - correlation_id
    - trace_id

3) "Free-form" event payloads are usually a bad idea, because it means that the developer is constantly having to remember to update the event, add attributes, and somehow populate them within the code. Instead I would suggest that event schemas are derived from "models" - each event just has a header, and a model. With a bit of code-generation we can then add a method to each model which prepares and fires an event. This way you know that whenever the models / data changes, your events will be updated too.

**Comment by chrusty** [comment]
This is an events interface I made for another client:
```go
package events

import (
	"github.com/golang/protobuf/proto"
)

// Publisher defines a standard events publisher interface (so we can plug in different implementations):
type Publisher interface {
	// Publish() instructs a publisher to immediately publish a message to the events topic:
	Publish(message proto.Message) error
}

// Subscriber defines a standard events subscriber interface (so we can plug in different implementations):
type Subscriber interface {
	// Firehose() instructs a subscriber to consume events from the general events/firehose topic, and return a channel which it will put them on:
	Firehose() (chan FirehoseEvent, error)
	// Subscribe() instructs a subscriber to consume events of a particular type, and return a channel which it will put them on:
	Subscribe(messageType proto.Message) (chan Event, error)
}

// Event defines a standard event interface for subscribers to provide:
type Event interface {
	// Ack() tells the subscriber that we have successfully processed this message:
	Ack()
	// Nack() tells the subscriber that we failed to process this message, please try to re-queue it:
	Nack()
	// Payload() returns the payload in its reconstructed form:
	Payload() proto.Message
	// Type() returns the event-type:
	Type() string
}

// FirehoseEvent defines a standard event interface for firehoses to provide:
type FirehoseEvent interface {
	// Ack() tells the subscriber that we have successfully processed this message:
	Ack()
	// Nack() tells the subscriber that we failed to process this message, please try to re-queue it:
	Nack()
	// RawPayload() returns the raw JSON / wire-format message body (which the user will have to marshal / assert):
	RawPayload() []byte
	// Type() returns the event-type:
	Type() string
}
```

**Comment by chrusty** [comment]
> Other things to consider
> 
>     * what storage is backing this? Feels like cockroach won't be the right thing
> 
>     * how does querying work? Ideally you want to do things like "show me all the events for this customer" , "show me all events for this namespace"

This very quickly becomes a big data problem. I'd suggest that whatever we pick should support TTLs, and make sure that we create aggregate datasets within the expiry time. Maybe the event-store should be a different ticket? The firehose / event-stream itself can run without having to store stuff initially and still provide value to us.

Keeping it ephemeral at first makes this bit much easier.

**Comment by domwong** [comment]
> Publish to one topic then have another service which reads all of those and splits them to their own topics.

I think my preference is for this approach, mostly because it means any particular service isn't responsible for knowing how to classify its own events. 

> Schema library is an absolute must

Agree. Do you have any feelings on where schema definitions live? In some central repository or with the service itself? Seems like with the service is better if we want events to be connected to their models per your following point.

> "Free-form" event payloads are usually a bad idea,

Makes sense

**Comment by chrusty** [comment]
I've usually done this with a shared "schema library". The problem with trying to attach them to services is that many of your domain objects won't belong to just one service. Matter of opinion really, but in mine this works best as a monorepo.

Oh did I mention that events should be versioned from day 1?

**Comment by crufter** [comment]
+1 for a schema library.

I have linked this many time before in case I bore some of you with it sorry: https://github.com/DefinitelyTyped/DefinitelyTyped

That link is basically a repository for all TypeScript types that were created for originally JavaScript libraries. If TS people managed to put all their types into 1 repo across a whole language surely we can do well with a shared schema library.
